Alerting: Fix Simple condition threshold inputs with negative values.

soniaAguilarPeiron opened the pull request:
**What is this feature?**

Currently when using the simple condition in the query and expression step (alert rule form), user cannot type `-` to use a negative number and use down arrow instead.

This PR fixes this issue.

**Why do we need this feature?**

Users should be able to type `-` char for negative numbers.

**Who is this feature for?**

Alerting users.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

soniaAguilarPeiron committed:
first attempt to fix the issue

soniaAguilarPeiron commented:
/deploy-to-hg

soniaAguilarPeiron committed:
fix prettier

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into alerting/allow-presssing-minus-in-simple-condition

gillesdemey commented on public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/SimpleCondition.tsx:
```diff
@@ -72,22 +72,43 @@ export const SimpleConditionEditor = ({
   const onEvaluateValueChange = (event: FormEvent<HTMLInputElement>, index?: number) => {
+    const value = event.currentTarget.value;
+
+    // Allow empty value
```
I think there's an opportunity to simplify a lot of the code here; a couple of questions about how it works so far and my take on it below

```
const onEvaluateValueChange = (event: FormEvent<HTMLInputElement>, index = 0) => {
    const value = event.currentTarget.value;
    const numericValue = parseFloat(value) || 0; // try to convert input to a number that isn't NaN

    onChange(
      produce(simpleCondition, (draftCondition) => {
        draftCondition.evaluator.params[index] = numericValue;
      })
    );

    updateThresholdValue(numericValue, index, expressionQueriesList, dispatch);
  };
  ```

this makes a few simplification and assumptions I wasn't quite sure about

1. do we _need_ to support `undefined` as the param? Can't we just assume the threshold is `0`?
2. it will always coerce the value to a numeric value, which I believe is correct?
3. removes some unnecessary spreading since we're already using immer anyway :)
4. no need to check for NaN if we use `|| 0` to coerce anything it can't parse to just `0`.

I wasn't quite sure why we needed `isRange` – I assume the `index` that is passed into the function will just place the param in the correct place in the array regardless of whether it's a range type or not

soniaAguilarPeiron replied:
Simplifying the isRange conditional works well though. I pushed a commit with this change.

gillesdemey replied:
I'm not entirely sure if I understand under which condition the `0` would need to be removed – is that for a range variant?

soniaAguilarPeiron replied:
what I mean is that when you have a 0 in the input, you cannot remove it if you are editing it 🤷‍♀️

gillesdemey commented:
Unrelated to the changes in this PR but I've noticed that we aren't showing two inputs for `IsOutsideRangeIncluded` and `isWithinRangeIncluded`.


```typescript
const isRange =
    simpleCondition.evaluator.type === EvalFunction.IsWithinRange ||
    simpleCondition.evaluator.type === EvalFunction.IsOutsideRange ||
    simpleCondition.evaluator.type === EvalFunction.IsOutsideRangeIncluded ||
    simpleCondition.evaluator.type === EvalFunction.IsWithinRangeIncluded;
```

I'll append those to the PR :)

gillesdemey committed:
Add additional condition types for included range functions

soniaAguilarPeiron committed:
simplify onEvaluateValueChange

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/SimpleCondition.tsx:
```diff
@@ -20,7 +20,7 @@ import { updateExpression } from './reducer';
   whenField?: string;
   evaluator: {
-    params: number[];
+    params: Array<number | string | undefined>;
```
Why we need to allow `string` and `undefined`?

soniaAguilarPeiron replied:
Because the code handles the case of typing a negative number:  when a user is typing a negative number, they first type the '-' character. During this intermediate state, the value is a string ('-') before it becomes a complete negative number.

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/SimpleCondition.tsx:
```diff
@@ -71,24 +73,37 @@ export const SimpleConditionEditor = ({
+    if (value === '') {
       const newParams = produce(simpleCondition.evaluator.params, (draft) => {
-        draft[index ?? 0] = parseFloat(event.currentTarget.value);
+        draft[index ?? 0] = undefined;
```
`index` defaults to `0` so it should never be undefined

soniaAguilarPeiron replied:
💯

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/SimpleCondition.tsx:
```diff
@@ -71,24 +73,37 @@ export const SimpleConditionEditor = ({
+    // Allow '-' for negative numbers
+    if (value === '-') {
+      const newParams = produce(simpleCondition.evaluator.params, (draft) => {
+        draft[index ?? 0] = '-';
```
This is quite confusing. We set `draft[index] = '-'`, so where we set the actual value?

soniaAguilarPeiron replied:
we set the actual value (in this case `-`, in `simpleCondition.evaluator.params[index]`

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/SimpleCondition.tsx:
```diff
@@ -71,24 +73,37 @@ export const SimpleConditionEditor = ({
+      return;
+    }
+    // Allow '-' for negative numbers
+    if (value === '-') {
```
Why we need to do anything special to parse negative values?

soniaAguilarPeiron replied:
this is the special case for the single '-' char

konrad147 commented on public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/SimpleCondition.tsx:
```diff
@@ -71,24 +73,37 @@ export const SimpleConditionEditor = ({
+      return;
+    }
+
+    const newParams = produce(simpleCondition.evaluator.params, (draft) => {
```
We already use immer here, so maybe it would be easier to do 
```suggestion
    const newParams = produce(simpleCondition, (draft) => {
      draft.evaluator.params[index] = numericValue;
      });
```
Then, we wouldn't need to create one more new object for the `onChange` below

soniaAguilarPeiron committed:
address feedback

soniaAguilarPeiron committed:
add onblur

soniaAguilarPeiron committed:
linter, translations

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into alerting/allow-presssing-minus-in-simple-condition

soniaAguilarPeiron committed:
betterer

konrad147 commented:
/deploy-to-hg

soniaAguilarPeiron commented:
/deploy-to-hg

soniaAguilarPeiron committed:
fix conflicts

soniaAguilarPeiron committed:
prettier

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into alerting/allow-presssing-minus-in-simple-condition

gillesdemey committed:
Refactor SimpleCondition component to use controlled inputs

Improves input handling for numeric values and fixes UI state issues
by using key prop to force re-render when values change.

gillesdemey committed:
Refactor range evaluator check into utility function

gillesdemey commented:
/deploy-to-hg

gillesdemey commented:
CI on Drone is failing because we're running an older version of `go` in this branch – I'll just keep an eye on the GitHub actions and merge if those are green.

gillesdemey committed:
Merge remote-tracking branch 'origin/main' into alerting/allow-presssing-minus-in-simple-condition

soniaAguilarPeiron committed:
fix betterer

soniaAguilarPeiron commented:
/deploy-to-hg
